### PR TITLE
[COLAB-2563] replace c-form bootstrap

### PR DESCRIPTION
**Pull request by cwoebker**
Using Bootstrap 4.0 form design with server-side verification. https://getbootstrap.com/docs/4.0/components/forms/#server-side

Still a work-in-progress. But as I am currently working through all the fields in the codebase, I figured it might make sense that one of you guys already take a look.

---
This change is [Reviewable](https://reviewable.io/reviews/cci-mit/xcolab/125)

### CODE REVIEW

**Comment by jobachhu** [comment]
Reviewed 1 of 1 files at r1.
Review status: all files reviewed at latest revision, all discussions resolved.

---

*[view/src/main/webapp/WEB-INF/jsp/profile/editUserProfile.jspx, line 78 at r1](https://beta.reviewable.io/reviews/cci-mit/xcolab/125#-L89hhIa7bTEe_0jghYT:-L89hhIa7bTEe_0jghYU:bhqob3w) ([raw file](https://github.com/cci-mit/xcolab/blob/d3c84819fabbdf6fc154d9b48ee80c1d77a4d214/view/src/main/webapp/WEB-INF/jsp/profile/editUserProfile.jspx#L78)):*
> ```
>                         <div class="col">
>                             <spring:bind path="firstName">
>                                 <form:input id="firstName" path="firstName" cssClass="form-control ${status.error ? 'is-invalid' : 'is-valid'}" autocomplete="given-name"/>
> ```

The `spring:bind` is a cool idea! I think we might actually be able to get around using it though, since Spring's form tags provide a means of defining different classes for valid and invalid states. You can use the `cssErrorClass="form-control is-invalid"` attribute on the input element for that.

One other thought is whether we want to set the `is-valid` class at all. We'd likely only want to set it after the page is submitted for the first time, otherwise all fields start out green. I think it's fine for the valid fields to just keep the neutral color, that makes it simpler. The invalid fields will stand out with their red color anyways. What do you think?

---


*Comments from [Reviewable](https://beta.reviewable.io/reviews/cci-mit/xcolab/125)*

**Comment by cwoebker** [comment]
Review status: 0 of 2 files reviewed at latest revision, 2 unresolved discussions.

---

*[view/src/main/webapp/WEB-INF/jsp/profile/editUserProfile.jspx, line 77 at r1](https://reviewable.io:443/reviews/cci-mit/xcolab/125#-L88hE6kxy4Y5_NgX5-A-r1-77:-L8_OTPyBoOfS1amEd9L:b-896fix) ([raw file](https://github.com/cci-mit/xcolab/blob/d3c84819fabbdf6fc154d9b48ee80c1d77a4d214/view/src/main/webapp/WEB-INF/jsp/profile/editUserProfile.jspx#L77)):*
<details><summary><i>Previously, cwoebker (Cecil Woebker) wrote…</i></summary><blockquote>

I am thinking of creating a custom component for the form:element + errors so reuse Some of this Stuff. That way we can change that more easily in the future.
</blockquote></details>

Done.

---

*[view/src/main/webapp/WEB-INF/jsp/profile/editUserProfile.jspx, line 78 at r1](https://reviewable.io:443/reviews/cci-mit/xcolab/125#-L89hhIa7bTEe_0jghYT:-L8_OR90clcE3yJ-CvgS:b-896fix) ([raw file](https://github.com/cci-mit/xcolab/blob/d3c84819fabbdf6fc154d9b48ee80c1d77a4d214/view/src/main/webapp/WEB-INF/jsp/profile/editUserProfile.jspx#L78)):*
<details><summary><i>Previously, jobachhu (Johannes Bachhuber) wrote…</i></summary><blockquote>

The `spring:bind` is a cool idea! I think we might actually be able to get around using it though, since Spring's form tags provide a means of defining different classes for valid and invalid states. You can use the `cssErrorClass="form-control is-invalid"` attribute on the input element for that.

One other thought is whether we want to set the `is-valid` class at all. We'd likely only want to set it after the page is submitted for the first time, otherwise all fields start out green. I think it's fine for the valid fields to just keep the neutral color, that makes it simpler. The invalid fields will stand out with their red color anyways. What do you think?
</blockquote></details>

Done.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cci-mit/xcolab/125)*

File: view/src/main/webapp/WEB-INF/jsp/profile/editUserProfile.jspx [hunk 1]
```diff
@@ -73,23 +74,21 @@
                             <spring:message code="profile.form.fields.firstname"/>
                         </div>
                         <div class="col">
-                            <form:input id="firstName" path="firstName" cssClass="form-control"
-                                        autocomplete="given-name"/>
-                            <div class="reg_errors"><!--  -->
```
[the remark this answers is not in the document]
Comment (cwoebker) [reply view/src/main/webapp/WEB-INF/jsp/profile/editUserProfile.jspx]: @jobachhu do you agree with this one? Or are those used elsewhere?

[the remark this answers is not in the document]
Comment (jobachhu) [reply view/src/main/webapp/WEB-INF/jsp/profile/editUserProfile.jspx]: @cwoebker I agree, we don't need that class anymore. It doesn't look like there are any CSS rules associated with it actually. It's only used once in JavaScript for some custom validation, but you can replace it with a better named, javascript-only class there (with `js-` prefix).

**Comment by jobachhu** [comment]
Reviewed 2 of 2 files at r2.
Review status: all files reviewed at latest revision, 2 unresolved discussions.

---

*[view/src/main/webapp/WEB-INF/tags/inputWithError.tagx, line 6 at r2](https://reviewable.io:443/reviews/cci-mit/xcolab/125#-L8bCpv_TI4AUhS4C-k9:-L8bCpv_TI4AUhS4C-kA:bh83pwa) ([raw file](https://github.com/cci-mit/xcolab/blob/a3359ca334e5edcaa0754213830d7b9fb4bed11b/view/src/main/webapp/WEB-INF/tags/inputWithError.tagx#L6)):*
> ```
> 
>     <jsp:directive.attribute name="title" type="java.lang.String" description="Title " />
>     <jsp:directive.attribute name="autocomplete" type="java.lang.String" description="Autocomplete " />
> ```

You might need some more attributes here later, e.g. I think we sometimes need other classes.

If it becomes a problem to do this with the JSP tag, let me know - then we can think about creating a Java-based tag (similar to `org.xcolab.view.tags.ImageTag`). That way it's easier to add more optional attributes.

---

*[view/src/main/webapp/WEB-INF/tags/inputWithError.tagx, line 8 at r2](https://reviewable.io:443/reviews/cci-mit/xcolab/125#-L8bDeEgOb9Bo-Kcy5bO:-L8bDeEgOb9Bo-Kcy5bP:b-4cggxz) ([raw file](https://github.com/cci-mit/xcolab/blob/a3359ca334e5edcaa0754213830d7b9fb4bed11b/view/src/main/webapp/WEB-INF/tags/inputWithError.tagx#L8)):*
> ```
>     <jsp:directive.attribute name="autocomplete" type="java.lang.String" description="Autocomplete " />
> 
>     <form:input id="${title}" path="${title}" cssClass="form-control" cssErrorClass="form-control is-invalid" autocomplete="${autocomplete}"/>
> ```

We might not always want the id to be equal to the title. I'd make this an optional attribute and default to the title.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cci-mit/xcolab/125)*

**Comment by jobachhu** [comment]
Reviewed 1 of 13 files at r3.
Review status: 1 of 13 files reviewed at latest revision, 1 unresolved discussion, some commit checks failed.

---

*[a discussion](https://beta.reviewable.io/reviews/cci-mit/xcolab/125#-L99z4vY3CIL0xttIHXs:-L99z4vY3CIL0xttIHXt:bs87y1v) (no related file):*
@cwoebker I'm not quite sure anymore if the tag is the right solution. The boilerplate they help us remove is quite small, and we need separate tags for each type of form field (you've already done `input` and `password`, but there are also `select` and `textarea` - so there's some repetition there as well). What do you think about not using a separate tag (then we also don't have to worry about redefining all attributes) and just adding the CSS error class and tag each time? I know there is some repetition, but that might be better than the additional complexity.

---

*[view/src/main/java/org/xcolab/view/tags/FormFieldTag.java, line 6 at r3](https://beta.reviewable.io/reviews/cci-mit/xcolab/125#-L99yemy4fs9p7ReD-z0:-L99yemy4fs9p7ReD-z1:btllhb) ([raw file](https://github.com/cci-mit/xcolab/blob/555963a8c5044f7a1c227b94bb1367cced904da4/view/src/main/java/org/xcolab/view/tags/FormFieldTag.java#L6)):*
> ```Java
> import javax.servlet.jsp.JspException;
> 
> public class FormFieldTag extends AbstractHtmlTag {
> ```

This still has the content of the ImageTag and seems to be unused - can we remove this or do you want to replace the JSP tags with this?

---

*[view/src/main/java/org/xcolab/view/tags/FormFieldTag.java, line 8 at r3](https://beta.reviewable.io/reviews/cci-mit/xcolab/125#-L99ymhx8E--iAfxX-g0:-L99ymhx8E--iAfxX-g1:b-26oy9y) ([raw file](https://github.com/cci-mit/xcolab/blob/555963a8c5044f7a1c227b94bb1367cced904da4/view/src/main/java/org/xcolab/view/tags/FormFieldTag.java#L8)):*
> ```Java
> public class FormFieldTag extends AbstractHtmlTag {
> 
>     private static final String FORM_FIELD_TAG = "form:input";
> ```

I'm not sure if you can just reference non-standard tags here and if they will be resolved.

---

*[view/src/main/webapp/WEB-INF/tags/inputWithError.tagx, line 12 at r3](https://beta.reviewable.io/reviews/cci-mit/xcolab/125#-L99yDTD7ZnXlSs0kNU-:-L99yDTD7ZnXlSs0kNU0:bfabykd) ([raw file](https://github.com/cci-mit/xcolab/blob/555963a8c5044f7a1c227b94bb1367cced904da4/view/src/main/webapp/WEB-INF/tags/inputWithError.tagx#L12)):*
> ```
>     <c:set var="id" value="${(empty id) ? path : id}" />
>     <c:set var="autocomplete" value="${(empty autocomplete) ? 'off' : autocomplete}" />
>     <c:set var="autocomplete" value="${(empty autocomplete) ? '' : disabled}" />
> ```

You're setting the same variable twice here.

---


*Comments from [Reviewable](https://beta.reviewable.io/reviews/cci-mit/xcolab/125)*

**Comment by cwoebker** [comment]
Review status: 4 of 21 files reviewed at latest revision, 6 unresolved discussions.

---

*[a discussion](https://beta.reviewable.io/reviews/cci-mit/xcolab/125#-L99z4vY3CIL0xttIHXs:-L9gzyRs3Jv43qEJ-WHi:b-ticrar) (no related file):*
<details><summary><i>Previously, jobachhu (Johannes Bachhuber) wrote…</i></summary><blockquote>

@cwoebker I'm not quite sure anymore if the tag is the right solution. The boilerplate they help us remove is quite small, and we need separate tags for each type of form field (you've already done `input` and `password`, but there are also `select` and `textarea` - so there's some repetition there as well). What do you think about not using a separate tag (then we also don't have to worry about redefining all attributes) and just adding the CSS error class and tag each time? I know there is some repetition, but that might be better than the additional complexity.
</blockquote></details>

Removed them.

---

*[view/src/main/webapp/WEB-INF/jsp/contestmanagement/details/translationTab.jspx, line 23 at r4](https://beta.reviewable.io/reviews/cci-mit/xcolab/125#-L9gapml8K-dUOLNfTYv:-L9gzoju-Cu8dEmIRqfc:b-896fix) ([raw file](https://github.com/cci-mit/xcolab/blob/196728534d2a36baba97524a195efae0a89cca6b/view/src/main/webapp/WEB-INF/jsp/contestmanagement/details/translationTab.jspx#L23)):*
<details><summary><i>Previously, jobachhu (Johannes Bachhuber) wrote…</i></summary><blockquote>

> ```
> cb-alert-form-invalid
> ```
I think we don't need this here, the alert can be a block element. We just needed this for the inline alerts for individual form fields (applies to the other cases as well).
</blockquote></details>

Done.

---

*[view/src/main/java/org/xcolab/view/tags/FormFieldTag.java, line 6 at r3](https://beta.reviewable.io/reviews/cci-mit/xcolab/125#-L99yemy4fs9p7ReD-z0:-L9gzs1w8koxuoWY4vQm:b-896fix) ([raw file](https://github.com/cci-mit/xcolab/blob/68d849552b690914ee3ea51d09cee5a862aa4b37/view/src/main/java/org/xcolab/view/tags/FormFieldTag.java#L6)):*
<details><summary><i>Previously, jobachhu (Johannes Bachhuber) wrote…</i></summary><blockquote>

This still has the content of the ImageTag and seems to be unused - can we remove this or do you want to replace the JSP tags with this?
</blockquote></details>

Done.

---

*[view/src/main/java/org/xcolab/view/tags/FormFieldTag.java, line 8 at r3](https://beta.reviewable.io/reviews/cci-mit/xcolab/125#-L99ymhx8E--iAfxX-g0:-L9gzsOqBddKs16UBF9e:b-896fix) ([raw file](https://github.com/cci-mit/xcolab/blob/68d849552b690914ee3ea51d09cee5a862aa4b37/view/src/main/java/org/xcolab/view/tags/FormFieldTag.java#L8)):*
<details><summary><i>Previously, jobachhu (Johannes Bachhuber) wrote…</i></summary><blockquote>

I'm not sure if you can just reference non-standard tags here and if they will be resolved.
</blockquote></details>

Done.

---

*[view/src/main/webapp/WEB-INF/tags/inputWithError.tagx, line 8 at r2](https://beta.reviewable.io/reviews/cci-mit/xcolab/125#-L8bDeEgOb9Bo-Kcy5bO:-L9gztR6Cs5aub_4FXYU:b-896fix) ([raw file](https://github.com/cci-mit/xcolab/blob/a3359ca334e5edcaa0754213830d7b9fb4bed11b/view/src/main/webapp/WEB-INF/tags/inputWithError.tagx#L8)):*
<details><summary><i>Previously, jobachhu (Johannes Bachhuber) wrote…</i></summary><blockquote>

We might not always want the id to be equal to the title. I'd make this an optional attribute and default to the title.
</blockquote></details>

Done.

---

*[view/src/main/webapp/WEB-INF/tags/inputWithError.tagx, line 12 at r3](https://beta.reviewable.io/reviews/cci-mit/xcolab/125#-L99yDTD7ZnXlSs0kNU-:-L9gztou4osYnQPZOWbE:b-896fix) ([raw file](https://github.com/cci-mit/xcolab/blob/68d849552b690914ee3ea51d09cee5a862aa4b37/view/src/main/webapp/WEB-INF/tags/inputWithError.tagx#L12)):*
<details><summary><i>Previously, jobachhu (Johannes Bachhuber) wrote…</i></summary><blockquote>

You're setting the same variable twice here.
</blockquote></details>

Done.

---


*Comments from [Reviewable](https://beta.reviewable.io/reviews/cci-mit/xcolab/125)*

**Comment by jobachhu** [comment]
:lgtm:

---

Reviewed 17 of 17 files at r5.
Review status: all files reviewed at latest revision, all discussions resolved.

---



*Comments from [Reviewable](https://beta.reviewable.io/reviews/cci-mit/xcolab/125#-:-L9jUdKW-xT1U3IA4mVB:bnfp4nl)*

**Pull request merged**